opentracing/README.md

```diff
@@ -0,0 +1,121 @@
+# DCP RFC
+
+For background:
+- [Main DCP doc](https://paper.dropbox.com/doc/Distributed-Context-Propagation-RGvlvD1NFKYmrJG9vGCES)
+- [DCP "Layers" doc](https://paper.dropbox.com/doc/DCP-Layers-and-Semantics-X1Tm1MSiBJECBkWYQKS2I)
+
+## API overview for those adding instrumentation
+
+Everyday consumers of this `opentracing` package really only need to worry
+about a couple of key abstractions: the `StartSpan` function, the `Span`
+interface, and golang's `context.Context` idiom. Here are code snippets
+demonstrating some important use cases.
+
+A note about `context.Context`: the opentracing API here encourages (but does
+not require) the use of golang-team's `context.Context` scheme. Whether we like
+it or not (this author is on the fence), `context.Context` is here to stay,
+goroutine-local-storage bedamned. When in Rome, do as the Romans do (etc).
+
+#### Singleton initialization
+
+The simplest starting point is `./global.go`. As early as possible, call
+
+    import ".../opentracing"
+    
+    func main() {
+        tracingRecorder := ... // tracing impl specific
+        tracingContextIDSource := ... // tracing impl specific
+        opentracing.InitGlobalTracer(tracingRecorder, tracingContextIDSource)
+        ...
+    }
+
+##### Note: the singletons are optional
+
+If global singletons make you sad, use `opentracing.NewStandardTracer(...)`
+directly and manage ownership of the `opentracing.OpenTracer` explicitly.
+
+#### Creating a Span given an existing Golang `context.Context`
+
+    func xyz(ctx context.Context, ...) {
+        ...
+        sp, ctx := opentracing.StartSpan("span_name", ctx)
+        defer sp.Finish()
+		sp.Info("called xyz")
+        ...
+    }
+
+#### Creating a Span given an existing Span
+
+    func xyz(parentSpan opentracing.Span, ...) {
+        ...
+        sp, ctx := opentracing.StartSpan("span_name", parentSpan)
+        defer sp.Finish()
+		sp.Info("called xyz")
+        ...
+    }
+
+#### Creating a root Span (i.e., without a known parent)
+
+    func xyz() {
+        ...
+        sp, ctx := opentracing.StartSpan("span_name")
+        defer sp.Finish()
+		sp.Info("called xyz")
+        ...
+    }
+
+#### Serializing to the wire
+
+    func makeSomeRequest(ctx context.Context) ... {
+        if span := SpanFromGoContext(ctx); span != nil {
+            httpClient := &http.Client{}
+            httpReq, _ := http.NewRequest("GET", "http://myservice/", nil)
+
+			// Transmit the span's ContextID as an HTTP header on our outbound
+            // request.
+            opentracing.AddContextIDToHttpHeader(span.ContextID(), httpReq.Header)
```

Comment:
I think we're missing an opportunity here for a more generic context propagation, a metadata context that contains arbitrary key/value attributes, of which tracing context ID is just one example. Security credentials can be another example. I just started working on such generic propagation API internally, so don't have any code to share, or even clear idea how it would work. The basic idea though is that the application can opt-in for different types of metadata propagation, e.g. tracing and security credentials, each of which register a handler with DCP that's invoked at the time when incoming or outgoing requests are made, and the handlers (de)serialize the metadata from the headers into metadata context. The point where I am undecided is whether the handlers should be invoked above or inside the spans manipulation, i.e. if it's above then the span manipulation is actually an implementation of the tracing handler. Opinions are welcome.